ruby/rspec/tutorial.md

```diff
@@ -0,0 +1,154 @@
+
+We use Capybara for interacting with the browser and PhantomJS (through the ruby driver poltergeist) or webkit to test pages with js. You can take a look at the capybara’s GitHub page [about drivers](https://github.com/teamcapybara/capybara#drivers). If your test doesn't need Javascript in order to verify the behavior, don't enable it. An example of feature test is:
+
+```
+describe 'User creates a task' do
+  before do
+    visit new_task_path
+  end
+
+  it 'can create a task' do
+    fill_in 'Title', with: 'My new task'
+    fill_in 'Description', with: 'I need to do the laundry'
+    click_button 'Create Task'
+
+    expect(page).to have_content 'The task was created'
+  end
+end
+```
+
+Also, Capybara comes with a built-in DSL for creating descriptive acceptance tests, where the previous code can be rewritten as:
+
+```
+feature 'User creates a task' do
+  background do
+    visit new_task_path
+  end
+
+  scenario 'can create a task' do
+    fill_in 'Title', with: 'My new task'
+    fill_in 'Description', with: 'I need to do the laundry'
+    click_button 'Create Task'
+
+    expect(page).to have_content 'The task was created'
+  end
+end
+```
+
+Both ways are accepted but the first one is the most used in the company because the other types of testings use the same structure. We recommend checking [this guide](https://robots.thoughtbot.com/write-reliable-asynchronous-integration-tests-with-capybara) that brings some tips at the time of doing feature testing.
+
+
+## Controller specs
+Although feature specs are great for acceptance tests, they are slower to run. So leave the edge case tests to your model or controller. However, don't use them for API test, you should be using requests specs for that. An example of it is the following:
+
+```
+describe RegistrationsController do  
+  describe 'POST create' do
+    let(:email) { 'email@example.com' }
+    let(:password) { 'somepassword' }
+
+    it 'redirects to the dashboard' do
+      post :create, { email: email, password: password }
+      expect(response).to redirect_to '/dashboard'
+    end
+
+    it 'creates an user' do
+      expect do
+        post :create, { email: email, password: password }
+      end.to change(User, :count).by(1)
+    end
+
+    it 'shows a successfully flash' do
+      post :create, { email: email, password: password }
+      expect(flash[:notice]).to match(/You can start using the app!/)
+    end    
+  end
+end
+```
+
+## Request specs
+Requests specs are the integration tests for APIs, they are designed to drive behavior through the full stack. We recommend doing requests specs over controller specs for APIs because the routes and real responses are covered. An example of them are:
+
```

Comment:
In addition, they are not so slow as capybara feature tests.

Reply:
I didn't mention capybara here but I can add what you said

Reply:
I would say browser feature tests or something like that.